### PR TITLE
tests: fix multiprocessing failures

### LINKED ISSUE #186: pytest sometimes throws `ConftestImportFailure` with pytest-xdist (multiprocessing) on macOS when testing wheel builds
```
$ tox -e py39-wheel
.package create: /Users/layday/code/build/.tox/.package
.package installdeps: setuptools >= 40.8.0, wheel
py39-wheel create: /Users/layday/code/build/.tox/py39-wheel
py39-wheel installdeps: -rrequirements-dev.txt
py39-wheel inst: /Users/layday/code/build/.tox/.tmp/package/1/build-0.1.0.tar.gz
py39-wheel installed: apipkg==1.5,attrs==20.3.0,build @ file:///Users/layday/code/build/.tox/.tmp/package/1/build-0.1.0.tar.gz,coverage==5.3,execnet==1.7.1,filelock==3.0.12,iniconfig==1.1.1,packaging==20.4,pep517==0.9.1,pluggy==0.13.1,py==1.9.0,pyparsing==2.4.7,pytest==6.1.2,pytest-cov==2.10.1,pytest-forked==1.3.0,pytest-mock==3.3.1,pytest-xdist==2.1.0,six==1.15.0,toml==0.10.2
py39-wheel run-test-pre: PYTHONHASHSEED='735881058'
py39-wheel run-test: commands[0] | pytest -rsx --cov --cov-config setup.cfg --cov-report=html:/Users/layday/code/build/.tox/py39-wheel/htmlcov --cov-context=test --cov-report=xml:/Users/layday/code/build/.tox/coverage.py39-wheel.xml tests -n auto
================================================= test session starts =================================================
platform darwin -- Python 3.9.0, pytest-6.1.2, py-1.9.0, pluggy-0.13.1
cachedir: .tox/py39-wheel/.pytest_cache
rootdir: /Users/layday/code/build, configfile: setup.cfg
plugins: xdist-2.1.0, mock-3.3.1, cov-2.10.1, forked-1.3.0
gw0 C / gw1 C / gw2 C / gw3 Cerror: [Errno 2] No such file or directory: 'build/bdist.macosx-10.6-x86_64/wheel/build-0.1.0-py3.9.egg-info'
[gw1] node down: Traceback (most recent call last):
  File "/Users/layday/code/build/.tox/py39-wheel/lib/python3.9/site-packages/_pytest/config/__init__.py", line 570, in _importconftest
    mod = import_path(conftestpath, mode=importmode)
  File "/Users/layday/code/build/.tox/py39-wheel/lib/python3.9/site-packages/_pytest/pathlib.py", line 517, in import_path
    importlib.import_module(module_name)
  File "/nix/store/qd9a9m03k6d4j3fvrflfynzn0yq7sdcg-python3-3.9.0/lib/python3.9/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1030, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1007, in _find_and_load
  File "<frozen importlib._bootstrap>", line 986, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 680, in _load_unlocked
  File "/Users/layday/code/build/.tox/py39-wheel/lib/python3.9/site-packages/_pytest/assertion/rewrite.py", line 171, in exec_module
    exec(co, module.__dict__)
  File "/Users/layday/code/build/tests/conftest.py", line 35, in <module>
    _setup()
  File "/Users/layday/code/build/tests/conftest.py", line 28, in _setup
    subprocess.check_output(cmd, env=env)
  File "/nix/store/qd9a9m03k6d4j3fvrflfynzn0yq7sdcg-python3-3.9.0/lib/python3.9/subprocess.py", line 420, in check_output
    return run(*popenargs, stdout=PIPE, timeout=timeout, check=True,
  File "/nix/store/qd9a9m03k6d4j3fvrflfynzn0yq7sdcg-python3-3.9.0/lib/python3.9/subprocess.py", line 524, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '['/Users/layday/code/build/.tox/py39-wheel/bin/python', '-m', 'build', '--wheel', '--no-isolation', '--outdir', '/tmp/tmph3yrfgek']' returned non-zero exit status 1.

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/layday/code/build/.tox/py39-wheel/lib/python3.9/site-packages/execnet/gateway_base.py", line 1084, in executetask
    do_exec(co, loc)  # noqa
  File "/Users/layday/code/build/.tox/py39-wheel/lib/python3.9/site-packages/xdist/remote.py", line 227, in <module>
    config = _prepareconfig(args, None)
  File "/Users/layday/code/build/.tox/py39-wheel/lib/python3.9/site-packages/_pytest/config/__init__.py", line 318, in _prepareconfig
    config = pluginmanager.hook.pytest_cmdline_parse(
  File "/Users/layday/code/build/.tox/py39-wheel/lib/python3.9/site-packages/pluggy/hooks.py", line 286, in __call__
    return self._hookexec(self, self.get_hookimpls(), kwargs)
  File "/Users/layday/code/build/.tox/py39-wheel/lib/python3.9/site-packages/pluggy/manager.py", line 93, in _hookexec
    return self._inner_hookexec(hook, methods, kwargs)
  File "/Users/layday/code/build/.tox/py39-wheel/lib/python3.9/site-packages/pluggy/manager.py", line 84, in <lambda>
    self._inner_hookexec = lambda hook, methods, kwargs: hook.multicall(
  File "/Users/layday/code/build/.tox/py39-wheel/lib/python3.9/site-packages/pluggy/callers.py", line 203, in _multicall
    gen.send(outcome)
  File "/Users/layday/code/build/.tox/py39-wheel/lib/python3.9/site-packages/_pytest/helpconfig.py", line 100, in pytest_cmdline_parse
    config = outcome.get_result()  # type: Config
  File "/Users/layday/code/build/.tox/py39-wheel/lib/python3.9/site-packages/pluggy/callers.py", line 80, in get_result
    raise ex[1].with_traceback(ex[2])
  File "/Users/layday/code/build/.tox/py39-wheel/lib/python3.9/site-packages/pluggy/callers.py", line 187, in _multicall
    res = hook_impl.function(*args)
  File "/Users/layday/code/build/.tox/py39-wheel/lib/python3.9/site-packages/_pytest/config/__init__.py", line 1003, in pytest_cmdline_parse
    self.parse(args)
  File "/Users/layday/code/build/.tox/py39-wheel/lib/python3.9/site-packages/_pytest/config/__init__.py", line 1280, in parse
    self._preparse(args, addopts=addopts)
  File "/Users/layday/code/build/.tox/py39-wheel/lib/python3.9/site-packages/_pytest/config/__init__.py", line 1186, in _preparse
    self.hook.pytest_load_initial_conftests(
  File "/Users/layday/code/build/.tox/py39-wheel/lib/python3.9/site-packages/pluggy/hooks.py", line 286, in __call__
    return self._hookexec(self, self.get_hookimpls(), kwargs)
  File "/Users/layday/code/build/.tox/py39-wheel/lib/python3.9/site-packages/pluggy/manager.py", line 93, in _hookexec
    return self._inner_hookexec(hook, methods, kwargs)
  File "/Users/layday/code/build/.tox/py39-wheel/lib/python3.9/site-packages/pluggy/manager.py", line 84, in <lambda>
    self._inner_hookexec = lambda hook, methods, kwargs: hook.multicall(
  File "/Users/layday/code/build/.tox/py39-wheel/lib/python3.9/site-packages/pluggy/callers.py", line 208, in _multicall
    return outcome.get_result()
  File "/Users/layday/code/build/.tox/py39-wheel/lib/python3.9/site-packages/pluggy/callers.py", line 80, in get_result
    raise ex[1].with_traceback(ex[2])
  File "/Users/layday/code/build/.tox/py39-wheel/lib/python3.9/site-packages/pluggy/callers.py", line 187, in _multicall
    res = hook_impl.function(*args)
  File "/Users/layday/code/build/.tox/py39-wheel/lib/python3.9/site-packages/_pytest/config/__init__.py", line 1070, in pytest_load_initial_conftests
    self.pluginmanager._set_initial_conftests(early_config.known_args_namespace)
  File "/Users/layday/code/build/.tox/py39-wheel/lib/python3.9/site-packages/_pytest/config/__init__.py", line 500, in _set_initial_conftests
    self._try_load_conftest(anchor, namespace.importmode)
  File "/Users/layday/code/build/.tox/py39-wheel/lib/python3.9/site-packages/_pytest/config/__init__.py", line 508, in _try_load_conftest
    self._getconftestmodules(anchor, importmode)
  File "/Users/layday/code/build/.tox/py39-wheel/lib/python3.9/site-packages/_pytest/config/__init__.py", line 536, in _getconftestmodules
    mod = self._importconftest(conftestpath, importmode)
  File "/Users/layday/code/build/.tox/py39-wheel/lib/python3.9/site-packages/_pytest/config/__init__.py", line 574, in _importconftest
    raise ConftestImportFailure(conftestpath, exc_info) from e
_pytest.config.ConftestImportFailure: CalledProcessError: Command '['/Users/layday/code/build/.tox/py39-wheel/bin/python', '-m', 'build', '--wheel', '--no-isolation', '--outdir', '/tmp/tmph3yrfgek']' returned non-zero exit status 1. (from /Users/layday/code/build/tests/conftest.py)


replacing crashed worker gw1
gw0 C / gw4 C / gw2 C / gw3 CTraceback (most recent call last):
  File "/Users/layday/code/build/.tox/py39-wheel/lib/python3.9/site-packages/pep517/_in_process.py", line 280, in <module>
    main()
  File "/Users/layday/code/build/.tox/py39-wheel/lib/python3.9/site-packages/pep517/_in_process.py", line 263, in main
    json_out['return_val'] = hook(**hook_input['kwargs'])
  File "/Users/layday/code/build/.tox/py39-wheel/lib/python3.9/site-packages/pep517/_in_process.py", line 204, in build_wheel
    return _build_backend().build_wheel(wheel_directory, config_settings,
  File "/Users/layday/code/build/.tox/py39-wheel/lib/python3.9/site-packages/setuptools/build_meta.py", line 216, in build_wheel
    return self._build_with_temp_dir(['bdist_wheel'], '.whl',
  File "/Users/layday/code/build/.tox/py39-wheel/lib/python3.9/site-packages/setuptools/build_meta.py", line 202, in _build_with_temp_dir
    self.run_setup()
  File "/Users/layday/code/build/.tox/py39-wheel/lib/python3.9/site-packages/setuptools/build_meta.py", line 145, in run_setup
    exec(compile(code, __file__, 'exec'), locals())
  File "setup.py", line 4, in <module>
    setup()
  File "/Users/layday/code/build/.tox/py39-wheel/lib/python3.9/site-packages/setuptools/__init__.py", line 153, in setup
    return distutils.core.setup(**attrs)
  File "/nix/store/qd9a9m03k6d4j3fvrflfynzn0yq7sdcg-python3-3.9.0/lib/python3.9/distutils/core.py", line 148, in setup
    dist.run_commands()
  File "/nix/store/qd9a9m03k6d4j3fvrflfynzn0yq7sdcg-python3-3.9.0/lib/python3.9/distutils/dist.py", line 966, in run_commands
    self.run_command(cmd)
  File "/nix/store/qd9a9m03k6d4j3fvrflfynzn0yq7sdcg-python3-3.9.0/lib/python3.9/distutils/dist.py", line 985, in run_command
    cmd_obj.run()
  File "/Users/layday/code/build/.tox/py39-wheel/lib/python3.9/site-packages/wheel/bdist_wheel.py", line 342, in run
    self.egg2dist(self.egginfo_dir, distinfo_dir)
  File "/Users/layday/code/build/.tox/py39-wheel/lib/python3.9/site-packages/wheel/bdist_wheel.py", line 445, in egg2dist
    raise ValueError(err)
ValueError: Egg metadata expected at build/bdist.macosx-10.6-x86_64/wheel/build-0.1.0-py3.9.egg-info but not found
[gw3] node down: Traceback (most recent call last):
  File "/Users/layday/code/build/.tox/py39-wheel/lib/python3.9/site-packages/_pytest/config/__init__.py", line 570, in _importconftest
    mod = import_path(conftestpath, mode=importmode)
  File "/Users/layday/code/build/.tox/py39-wheel/lib/python3.9/site-packages/_pytest/pathlib.py", line 517, in import_path
    importlib.import_module(module_name)
  File "/nix/store/qd9a9m03k6d4j3fvrflfynzn0yq7sdcg-python3-3.9.0/lib/python3.9/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1030, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1007, in _find_and_load
  File "<frozen importlib._bootstrap>", line 986, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 680, in _load_unlocked
  File "/Users/layday/code/build/.tox/py39-wheel/lib/python3.9/site-packages/_pytest/assertion/rewrite.py", line 171, in exec_module
    exec(co, module.__dict__)
  File "/Users/layday/code/build/tests/conftest.py", line 35, in <module>
    _setup()
  File "/Users/layday/code/build/tests/conftest.py", line 28, in _setup
    subprocess.check_output(cmd, env=env)
  File "/nix/store/qd9a9m03k6d4j3fvrflfynzn0yq7sdcg-python3-3.9.0/lib/python3.9/subprocess.py", line 420, in check_output
    return run(*popenargs, stdout=PIPE, timeout=timeout, check=True,
  File "/nix/store/qd9a9m03k6d4j3fvrflfynzn0yq7sdcg-python3-3.9.0/lib/python3.9/subprocess.py", line 524, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '['/Users/layday/code/build/.tox/py39-wheel/bin/python', '-m', 'build', '--wheel', '--no-isolation', '--outdir', '/tmp/tmpu64dcw8m']' returned non-zero exit status 1.

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/layday/code/build/.tox/py39-wheel/lib/python3.9/site-packages/execnet/gateway_base.py", line 1084, in executetask
    do_exec(co, loc)  # noqa
  File "/Users/layday/code/build/.tox/py39-wheel/lib/python3.9/site-packages/xdist/remote.py", line 227, in <module>
    config = _prepareconfig(args, None)
  File "/Users/layday/code/build/.tox/py39-wheel/lib/python3.9/site-packages/_pytest/config/__init__.py", line 318, in _prepareconfig
    config = pluginmanager.hook.pytest_cmdline_parse(
  File "/Users/layday/code/build/.tox/py39-wheel/lib/python3.9/site-packages/pluggy/hooks.py", line 286, in __call__
    return self._hookexec(self, self.get_hookimpls(), kwargs)
  File "/Users/layday/code/build/.tox/py39-wheel/lib/python3.9/site-packages/pluggy/manager.py", line 93, in _hookexec
    return self._inner_hookexec(hook, methods, kwargs)
  File "/Users/layday/code/build/.tox/py39-wheel/lib/python3.9/site-packages/pluggy/manager.py", line 84, in <lambda>
    self._inner_hookexec = lambda hook, methods, kwargs: hook.multicall(
  File "/Users/layday/code/build/.tox/py39-wheel/lib/python3.9/site-packages/pluggy/callers.py", line 203, in _multicall
    gen.send(outcome)
  File "/Users/layday/code/build/.tox/py39-wheel/lib/python3.9/site-packages/_pytest/helpconfig.py", line 100, in pytest_cmdline_parse
    config = outcome.get_result()  # type: Config
  File "/Users/layday/code/build/.tox/py39-wheel/lib/python3.9/site-packages/pluggy/callers.py", line 80, in get_result
    raise ex[1].with_traceback(ex[2])
  File "/Users/layday/code/build/.tox/py39-wheel/lib/python3.9/site-packages/pluggy/callers.py", line 187, in _multicall
    res = hook_impl.function(*args)
  File "/Users/layday/code/build/.tox/py39-wheel/lib/python3.9/site-packages/_pytest/config/__init__.py", line 1003, in pytest_cmdline_parse
    self.parse(args)
  File "/Users/layday/code/build/.tox/py39-wheel/lib/python3.9/site-packages/_pytest/config/__init__.py", line 1280, in parse
    self._preparse(args, addopts=addopts)
  File "/Users/layday/code/build/.tox/py39-wheel/lib/python3.9/site-packages/_pytest/config/__init__.py", line 1186, in _preparse
    self.hook.pytest_load_initial_conftests(
  File "/Users/layday/code/build/.tox/py39-wheel/lib/python3.9/site-packages/pluggy/hooks.py", line 286, in __call__
    return self._hookexec(self, self.get_hookimpls(), kwargs)
  File "/Users/layday/code/build/.tox/py39-wheel/lib/python3.9/site-packages/pluggy/manager.py", line 93, in _hookexec
    return self._inner_hookexec(hook, methods, kwargs)
  File "/Users/layday/code/build/.tox/py39-wheel/lib/python3.9/site-packages/pluggy/manager.py", line 84, in <lambda>
    self._inner_hookexec = lambda hook, methods, kwargs: hook.multicall(
  File "/Users/layday/code/build/.tox/py39-wheel/lib/python3.9/site-packages/pluggy/callers.py", line 208, in _multicall
    return outcome.get_result()
  File "/Users/layday/code/build/.tox/py39-wheel/lib/python3.9/site-packages/pluggy/callers.py", line 80, in get_result
    raise ex[1].with_traceback(ex[2])
  File "/Users/layday/code/build/.tox/py39-wheel/lib/python3.9/site-packages/pluggy/callers.py", line 187, in _multicall
    res = hook_impl.function(*args)
  File "/Users/layday/code/build/.tox/py39-wheel/lib/python3.9/site-packages/_pytest/config/__init__.py", line 1070, in pytest_load_initial_conftests
    self.pluginmanager._set_initial_conftests(early_config.known_args_namespace)
  File "/Users/layday/code/build/.tox/py39-wheel/lib/python3.9/site-packages/_pytest/config/__init__.py", line 500, in _set_initial_conftests
    self._try_load_conftest(anchor, namespace.importmode)
  File "/Users/layday/code/build/.tox/py39-wheel/lib/python3.9/site-packages/_pytest/config/__init__.py", line 508, in _try_load_conftest
    self._getconftestmodules(anchor, importmode)
  File "/Users/layday/code/build/.tox/py39-wheel/lib/python3.9/site-packages/_pytest/config/__init__.py", line 536, in _getconftestmodules
    mod = self._importconftest(conftestpath, importmode)
  File "/Users/layday/code/build/.tox/py39-wheel/lib/python3.9/site-packages/_pytest/config/__init__.py", line 574, in _importconftest
    raise ConftestImportFailure(conftestpath, exc_info) from e
_pytest.config.ConftestImportFailure: CalledProcessError: Command '['/Users/layday/code/build/.tox/py39-wheel/bin/python', '-m', 'build', '--wheel', '--no-isolation', '--outdir', '/tmp/tmpu64dcw8m']' returned non-zero exit status 1. (from /Users/layday/code/build/tests/conftest.py)


replacing crashed worker gw3
gw0 [94] / gw4 [94] / gw2 [94] / gw5 [94]
.....................................................s.sss.sssssssssssssss.sssssssssss.s......                  [100%]

---------- coverage: platform darwin, python 3.9.0-final-0 -----------
---------------------- coverage: failed workers ----------------------
The following workers failed to return coverage data, ensure that pytest-cov is installed on these workers.
gw1
gw3
Coverage HTML written to dir /Users/layday/code/build/.tox/py39-wheel/htmlcov
Coverage XML written to file /Users/layday/code/build/.tox/coverage.py39-wheel.xml

=============================================== short test summary info ===============================================
SKIPPED [30] tests/test_integration.py:75: integration tests not run (no --run-integration flag)
SKIPPED [1] tests/test_integration.py: integration tests not run (no --run-integration flag)
=========================================== 63 passed, 31 skipped in 11.59s ===========================================
Requirement already satisfied: build==0.1.0 from file:///tmp/tmpruiwi01g/build-0.1.0-py2.py3-none-any.whl in ./.tox/py39-wheel/lib/python3.9/site-packages (0.1.0)
Requirement already satisfied: toml in ./.tox/py39-wheel/lib/python3.9/site-packages (from build==0.1.0) (0.10.2)
Requirement already satisfied: pep517>=0.9 in ./.tox/py39-wheel/lib/python3.9/site-packages (from build==0.1.0) (0.9.1)
Requirement already satisfied: packaging in ./.tox/py39-wheel/lib/python3.9/site-packages (from build==0.1.0) (20.4)
Requirement already satisfied: pyparsing>=2.0.2 in ./.tox/py39-wheel/lib/python3.9/site-packages (from packaging->build==0.1.0) (2.4.7)
Requirement already satisfied: six in ./.tox/py39-wheel/lib/python3.9/site-packages (from packaging->build==0.1.0) (1.15.0)
_______________________________________________________ summary _______________________________________________________
  py39-wheel: commands succeeded
  congratulations :)
```

The ultimate culprit appears to be the wheel package (or setuptools, whichever sets the output dir) which generates build artefacts in 'build' in the project root irrespective of the `outdir`.  This can give rise to several different race conditions, e.g. the metadata folder is deleted by another runner before it can be read, which is, presumably, what happened above:

```
[Errno 2] No such file or directory: 'build/bdist.macosx-10.6-x86_64/wheel/build-0.1.0-py3.9.egg-info'
```

And again:

```
ValueError: Egg metadata expected at build/bdist.macosx-10.6-x86_64/wheel/build-0.1.0-py3.9.egg-info but not found
```

No idea why the tests don't crash on other platforms.

### PR DESCRIPTION
**Pull request by layday**
The package would've been built simultaneously by each pytest-xdist
worker which would result in TOCTOUs (or something of the sort)
because temporary build artefacts were deposited in the same folder.

Closes #186.

### CODE REVIEW

**Review by gaborbernat — APPROVED** [review]

**Comment by layday** [comment]
erm, on second thought, I don't think it actually works this way -`build` has already been imported when the fixture is run.

**Comment by layday** [comment]
Okay, installing `build` happens before test collection again.  Please re-review.

File: tests/conftest.py [hunk 1]
```diff
@@ -12,27 +12,51 @@
 
 import pytest
 
+from filelock import FileLock
+
+
+def _build_and_reinstall_build(test_mode):
+    # print('building and reinstalling build {}'.format(test_mode), file=sys.stderr)
```
Comment (gaborbernat) [inline tests/conftest.py:19]: Remove this?

File: tests/conftest.py [hunk 2]
```diff
@@ -12,27 +12,51 @@
...
+            [sys.executable, '-m', 'build', '--{}'.format(test_mode), '--no-isolation', '--outdir', temp],
+        )
+        dist_file = next(t for t in os.listdir(temp) if t.endswith('.whl' if test_mode == 'wheel' else '.tar.gz'))
+        subprocess.check_call(
+            [
+                sys.executable,
+                '-m',
+                'pip',
+                'install',
+                # ``--upgrade`` will uninstall build prior to installing the build ``dist_file``
```
Comment (gaborbernat) [inline tests/conftest.py:32]: can this be on the right side of the ``--upgrade``?

**Review by gaborbernat — APPROVED** [review]

**Pull request merged**